### PR TITLE
Support for building with OpenJDK 21

### LINKED ISSUE #28293: ShardingSphere fails to build on OpenJDK 21 EA
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?

- https://github.com/apache/shardingsphere/commit/4563afd14d23d494f75f680c47a351cdd19893db

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

- `ShardingSphere-JDBC`

### Expected behavior

- Both Adoptium and GraalVM CE have created downstream builds of OpenJDK 21, time to validate OpenJDK 21 on the ShardingSphere side.  Refer to https://adoptium.net/zh-CN/blog/2023/08/early-access-builds/ and https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/23.1.0-dev-20230817_2024 . Considering that GraalVM CE 23.1.0 only have OpenJDK21 binary, it is very suitable to intervene in the investigation of ShardingSphere Proxy Native in advance.  Refer to https://github.com/apache/shardingsphere/issues/21347 .

- I created a small test at https://github.com/linghengqian/shardingsphere-jdk21-ea-ci-test and the results are at https://github.com/linghengqian/shardingsphere-jdk21-ea-ci-test/actions/runs/6003591784/job/16282365479 . Assume there are no errors reported for this CI.

### Actual behavior

-  It looks like some modules about `com.sun.tools.javac.tree.JCTree$JCImport` broke ShardingSphere builds on OpenJDK 21 EA.
 Refer to https://github.com/projectlombok/lombok/issues/3393 .
```shell
Error: ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project shardingsphere-test-util: Fatal error compiling: java.lang.NoSuchFieldError: Class com.sun.tools.javac.tree.JCTree$JCImport does not have member field 'com.sun.tools.javac.tree.JCTree qualid' -> [Help 1] 

org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project shardingsphere-test-util: Fatal error compiling 

at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:375)
```

### Reason analyze (If you can)

- Null.

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

- Switch to any OpenJDK 21 downstream distribution. Execute the following command.

```shell
git clone -b master https://github.com/apache/shardingsphere.git
cd ./shardingsphere/
git reset --hard 4563afd14d23d494f75f680c47a351cdd19893db
./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e
```

### Example codes for reproduce this issue (such as a github link).

- https://github.com/linghengqian/shardingsphere-jdk21-ea-ci-test

### PR DESCRIPTION
**Pull request by linghengqian**
Fixes #28293 .

Changes proposed in this pull request:
  - Support for building with OpenJDK 21.
  - Due to JEP 451, JDK21's CI has thousands more lines of warnings.  Reference https://github.com/mockito/mockito/issues/3037 .
  - Since GraalVM Updater is deprecated in GraalVM CE 23.1.0 For JDK 21, the `shardingsphere-infra-expr-espresso` submodule will not be reusable, there are some temporary processing at https://github.com/apache/shardingsphere/pull/28340 .  Looking to the future, it is impossible for us to introduce the GPL LICENSE Espresso ontology into ShardingSphere. This module will be deleted or maintained by a third party in the foreseeable future.
  - The javac in GraalVM CE 23.1.0 For JDK21 may has bugs. An error occurred when the ShardingSphere Proxy Native was migrated to GraalVM CE 23.1.0 For JDK21 build. The build of ShardingSphere Proxy Native will continue using GraalVM 23.0.1 For JDK17.
```bash
========================================================================================================================
GraalVM Native Image: Generating 'apache-shardingsphere-proxy-native' (executable)...
========================================================================================================================
[1/8] Initializing...                                                                                   (12.0s @ 0.41GB)
Java version: 21+35, vendor version: GraalVM CE 21+35.1
Graal compiler: optimization level: 2, target machine: x86-64-v3
C compiler: gcc (linux, x86_64, 11.4.0)
Garbage collector: Serial GC (max heap size: 80% of RAM)
2 user-specific feature(s):
- com.oracle.svm.polyglot.groovy.GroovyIndyInterfaceFeature
- com.oracle.svm.thirdparty.gson.GsonFeature
------------------------------------------------------------------------------------------------------------------------
Build resources:
- 9.57GB of memory (75.6% of 12.67GB system memory, determined at start)
- 6 thread(s) (100.0% of 6 available processor(s), determined at start)
[2/8] Performing analysis...  []                                                                        (69.2s @ 2.78GB)
27,510 reachable types   (76.6% of   35,901 total)
45,905 reachable fields  (52.2% of   87,956 total)
140,093 reachable methods (55.7% of  251,477 total)
8,227 types,   300 fields, and 6,865 methods registered for reflection

Error: java.util.concurrent.ExecutionException: java.lang.ClassCastException: class jdk.vm.ci.meta.NullConstant cannot be cast to class org.graalvm.compiler.core.common.type.TypedConstant (jdk.vm.ci.meta.NullConstant is in module jdk.internal.vm.ci of loader 'bootstrap'; org.graalvm.compiler.core.common.type.TypedConstant is in module jdk.internal.vm.compiler of loader 'platform')
------------------------------------------------------------------------------------------------------------------------
11.9s (14.3% of total time) in 55 GCs | Peak RSS: 6.23GB | CPU load: 5.30
========================================================================================================================
Finished generating 'apache-shardingsphere-proxy-native' in 1m 21s.
com.oracle.svm.driver.NativeImage$NativeImageError
at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.showError(NativeImage.java:2336)
at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.build(NativeImage.java:1919)
at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1878)
at org.graalvm.nativeimage.driver/com.oracle.svm.driver.NativeImage.main(NativeImage.java:1852)
at java.base@21/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
```

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [x] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: .github/workflows/nightly-ci.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ ubuntu-latest, macos-latest, windows-latest ]
-        java-version: [ 11, 17, 20 ]
+        java-version: [ 11, 21-ea ]
```
Comment (taojintianxia) [inline .github/workflows/nightly-ci.yml:44]: is it possible to keep the JDK 17 in matrix ?

File: pom.xml [hunk 2]
```diff
@@ -1235,7 +1256,6 @@
                             <excludes>
                                 <exclude>org.apache.shardingsphere.agent.core.**</exclude>
                             </excludes>
-                            <argLine>--add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.lang.reflect=ALL-UNNAMED</argLine>
```
Comment (taojintianxia) [inline pom.xml:1238]: what will happen if execute on JDK 8 or 17 without this argument ?

File: .github/workflows/nightly-ci.yml (hunk 1 above)
Comment (linghengqian) [reply .github/workflows/nightly-ci.yml]: - I've added back JDK17.

File: pom.xml (hunk 2 above)
Comment (linghengqian) [reply pom.xml:1238]: - It is impossible for the Maven Profile of `generateStandardMetadata` to run on JDK8.  This Maven Profile is restricted to run on JDK17-JDK21 through `maven-compiler-plugin`.  

- In addition, the Maven Profile of `jdk11+` also defines the same.

**Review by taojintianxia — APPROVED** [review]

**Pull request merged**